platform/android/java/lib/src/org/godotengine/godot/tts/GodotUtteranceListener.java

```diff
@@ -0,0 +1,131 @@
+/*************************************************************************/
+
+package org.godotengine.godot.tts;
+
+import org.godotengine.godot.*;
+
+import android.os.Bundle;
+import android.speech.tts.TextToSpeech;
+import android.speech.tts.UtteranceProgressListener;
+import android.speech.tts.Voice;
+
+import java.util.Iterator;
+import java.util.Set;
+
+class GodotUtteranceListener extends UtteranceProgressListener {
+	GodotTTS tts = null;
+
+	public void updateTTS() {
+		if (!tts.tts_speaking && tts.tts_queue.size() > 0) {
+			int mode = TextToSpeech.QUEUE_FLUSH;
+			GodotUtterance message = tts.tts_queue.getFirst();
+
+			Set<Voice> voices = tts.tts_synth.getVoices();
+			Iterator<Voice> it = voices.iterator();
+			while (it.hasNext()) {
+				Voice v = it.next();
+				if (v.getName() == message.voice) {
+					tts.tts_synth.setVoice(v);
+					break;
+				}
+			}
+			tts.tts_synth.setPitch(message.pitch);
+			tts.tts_synth.setSpeechRate(message.rate);
+
+			Bundle params = new Bundle();
+			params.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, message.volume / 100.f);
+
+			tts.last_utterance = message;
+			tts.last_utterance.start = 0;
+			tts.last_utterance.offset = 0;
+			tts.tts_paused = false;
+
+			tts.tts_synth.speak(message.text, mode, params, String.valueOf(message.id));
+			tts.tts_speaking = true;
+			tts.tts_queue.pollFirst();
+		}
+	}
+
+	@Override
+	public void onRangeStart(String utteranceId, int start, int end, int frame) {
+		if (Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			tts.last_utterance.offset = start;
+			GodotLib.ttsCallback(3 /*BOUNDARY*/, tts.last_utterance.id, start + tts.last_utterance.start);
+		}
+	}
+
+	@Override
+	public void onStop(String utteranceId, boolean interrupted) {
+		if (!tts.tts_paused && Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			GodotLib.ttsCallback(2 /*CANCEL*/, tts.last_utterance.id, 0);
+			tts.tts_speaking = false;
+			updateTTS();
+		}
+	}
+
+	@Override
+	public void onStart(String utteranceId) {
+		if (tts.last_utterance.start == 0 && Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			GodotLib.ttsCallback(0 /*START*/, tts.last_utterance.id, 0);
+		}
+	}
+
+	@Override
+	public void onDone(String utteranceId) {
+		if (!tts.tts_paused && Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			GodotLib.ttsCallback(1 /*END*/, tts.last_utterance.id, 0);
+			tts.tts_speaking = false;
+			updateTTS();
+		}
+	}
+
+	@Override
+	public void onError(String utteranceId) {
+		if (!tts.tts_paused && Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			GodotLib.ttsCallback(2 /*CANCEL*/, tts.last_utterance.id, 0);
+			tts.tts_speaking = false;
+			updateTTS();
+		}
+	}
+
+	@Override
+	public void onError(String utteranceId, int errorCode) {
+		if (!tts.tts_paused && Integer.parseInt(utteranceId) == tts.last_utterance.id) {
+			GodotLib.ttsCallback(2 /*CANCEL*/, tts.last_utterance.id, 0);
+			tts.tts_speaking = false;
+			updateTTS();
+		}
+	}
```

Comment:
So a couple of things:
- Can you create constants to use for the first `ttsCallback` argument.
- It seems like most of the logic here can easily be handled by `GodotTTS` instead. It should be fine to have `GodotTTS` extends `UtteranceProgressListener` instead of this class, and merge this logic there.